src/DynamicData.Tests/Cache/InnerJoinFixtureRaceCondition.cs

```diff
@@ -32,6 +32,25 @@ public void LetsSeeWhetherWeCanRandomlyHitARaceCondition()
         ids.InnerJoin(itemsCache.Connect(), x => x.Id, (_, thing) => thing).Subscribe((z) => { }, ex => { }, () => { });
     }
 
+    // See https://github.com/reactivemarbles/DynamicData/issues/787 
+    [Fact(Skip = "Heavyweight test so probably best run manually if this issue comes up again")]
+    public void LetsSeeWhetherWeCanRandomlyHitADifferentRaceCondition()
+    {
+        var leftCache = new SourceCache<Thing, long>(x => x.Id);
+        var rightCache = new SourceCache<Thing, long>(x => x.Id);
+        var joined = leftCache.Connect().InnerJoin(rightCache.Connect(), x => x.Id,
+            (keys, leftThing, rightThing) =>
+                new Thing { Id = keys.leftKey, Name = $"{leftThing.Name} x {rightThing.Name}" });
+
+        IDisposable StartUpdating(ISourceCache<Thing, long> sourceCache) =>
+            Observable.Range(1, 1_000_000, Scheduler.Default)
```

Comment:
Surely we could replicate the issue with fewer than 1,000,000 items. Then maybe the test would be feasible to include for normal runs? Is the issue really that rare?

Reply:
Perhaps we could make it 1000 times.  However, it may potentially become a traffic light test whereas with these numbers I was able to replicate it every time.

Reply:
Dropping to `100_000` gave me 10 failures in a row without the fix, and an average runtime of 500ms with the fix.

Reply:
I got it down to about 150ms on average with the fix, and 100 failures in a row without it.

I realized that there are 3 things racing against each other here, but the test was only repeating 2 of those things. The `.Subscribe()` call is the third thing, and it was only running once, right at the end of the test.

I also setup the item generator "threads" to run infinitely, but start removing items after the first 100, so that calls to `.Subscribe()` don't take progressively more and more time.

```cs
using var leftSource = new SourceCache<Thing, long>(thing => thing.Id);
using var rightSource = new SourceCache<Thing, long>(thing => thing.Id);

var resultStream = ObservableCacheEx.InnerJoin(
    left: leftSource.Connect(),
    right: rightSource.Connect(),
    rightKeySelector: rightThing => rightThing.Id,
    (keys, leftThing, rightThing) => new Thing()
    {
        Id = keys.leftKey,
        Name = $"{leftThing.Name} x {rightThing.Name}"
    });

using var leftThingGenerator = BeginGeneratingThings(leftSource, "Left");
using var rightThingGenerator = BeginGeneratingThings(rightSource, "Left");

for (var i = 0; i < 100; ++i)
{
    using var subscription = resultStream.Subscribe();
}

IDisposable BeginGeneratingThings(SourceCache<Thing, long> source, string namePrefix)
    => Observable.Range(1, int.MaxValue, ThreadPoolScheduler.Instance)
        .Subscribe(id =>
        {
            source.AddOrUpdate(new Thing()
            {
                Id = id,
                Name = $"{namePrefix}Thing #{id}"
            });
            source.RemoveKey(id - 100);
        });
```

Reply:
@JakenVeina you test is included in the PR now -> nice work.